draco/asp/hard.lp

```diff
@@ -1,24 +1,68 @@
 % ====== Hard constraints ======
 
-% @hard(text_mark_without_text_channel) Text mark requires text encoding.
-violation(text_mark_without_text_channel) :-
-    attribute((mark,type),M,text),
-    not attribute(mark_channel,M,text).
+% @hard(enc_type_valid) Primitive type has to support data type.
+violation(enc_type_valid) :-
+    attribute((field,type),F,(string;boolean)),
+    attribute((encoding,field),E,F),
+    attribute((encoding,type),E,(linear;log)).
 
-% @hard(text_channel_without_text_mark) Text channel requires text mark.
-violation(text_channel_without_text_mark) :-
-    attribute(mark_channel,M,text),
-    not attribute((mark,type),M,text).
+% @hard(log_non_positive) Cannot use log if the data is negative or zero.
+violation(log_non_positive) :-
+    attribute(extent,F,MIN,_),
+    attribute((encoding,field),E,F),
+    attribute((encoding,type),E,log),
+    MIN <= 0.
 
 % @hard(bin_and_aggregate) Cannot bin and aggregate.
 violation(bin_and_aggregate) :-
     attribute((encoding,binning),E,_),
     attribute((encoding,aggregate),E,_).
 
-% @hard(no_encodings) There has to be at least one encoding for every mark.
-violation(no_encodings) :-
-    entity(mark,_,M),
-    not entity(encoding,M,_).
+% @hard(aggregate_o_valid) Oridnal only supports min, max, and median.
+violation(aggregate_o_valid) :-
+    attribute((encoding,aggregate),E,A),
+    attribute((encoding,type),E,(ordinal;categorical)),
+    A != min,
+    A != max,
+    A != median.
+
+% @hard(aggregate_t_valid) Temporal only supports min and max.
+violation(aggregate_t_valid) :-
+    attribute((field,type),F,datetime),
+    attribute((encoding,field),E,F),
+    attribute((encoding,aggregate),E,A),
+    A != min,
+    A != max.
+
+% @hard(aggregate_detail) Detail cannot be aggregated.
+violation(aggregate_detail) :-
+    attribute((encoding,channel),E,detail),
+    attribute((encoding,aggregate),E,_).
+
+% @hard(count_without_field) Count cannot use a field.
+violation(count_without_field) :-
+    attribute((encoding,field),E,_),
+    attribute((encoding,aggregate),E,count).
+
+% @hard(count_without_q) Count has to have a quantitative scale.
+violation(count_without_q) :-
+    attribute((encoding,aggregate),E,count),
+    attribute((encoding,type),E,T),
+    T != log,
+    T != linear.
```

Comment:
We should probably add some fact for all the continuous scales. Make it similar to `non_positional` etc that we use for channel.